tchannel/tornado/connection.py

```diff
@@ -533,9 +533,11 @@ def send_error(self, error):
         error_message = build_raw_error_message(error)
         write_future = self.writer.put(error_message)
         write_future.add_done_callback(
-            lambda f: self.tchannel.event_emitter.fire(
-                EventType.after_send_error,
-                error,
+            lambda f: IOLoop.current().add_callback(
```

Comment:
I think we might still have an issue here, it will still execute the coroutine as if it were a regular function (thus possibly creating a race condition), but it's proooooobably fine since all the after_send_error does is log some stats, and I think we have some guarantee that the code will be at least executed.  Please correct me if I'm wrong on that.
cc @abhinav, @Groxx

Reply:
pretty sure it won't work as-is: http://www.tornadoweb.org/en/stable/ioloop.html#tornado.ioloop.IOLoop.add_callback
should be something like `add_callback(self.tchannel.event_emitter.fire, EventType.after_send_error, error)`

but yeah, afaik this is perfectly safe.  even if you don't explicitly yield it / put it in the ioloop, it still *runs* with the same guarantees as everything else (minus the fact that you've given up ordering), it just complains in logs that you never retrieved the value / exception.

Reply:
Actually the future does get resolved. `add_callback` causes the function to be called in the next loop, using `_run_callback`. `_run_callback` calls the function and does a `add_future` if the call results in a future (as this one will).

Ref: https://github.com/tornadoweb/tornado/blob/d68abdb4b68632f675f41ff817c370dbc8f1e21f/tornado/ioloop.py#L659

Reply:
Ah, derp, yeah - I just realized that's what he meant.  Thanks!

Reply:
As far as "send eventually"... given Writer's implementation (via Queue), it may not even be a behavioral change, aside from being slightly later (while still async).  I'm not familiar enough with the code to be sure about that though :|

Pragmatically, it doesn't look like you can / should expect to be able to change the error in the hook, so I don't think this makes a practical difference (especially since I think it's currently async too).

Seems fine.  Probably worth documenting if we think it does change the behavior, but anything it breaks seems like it'd have to be an un-supported use-case.

Reply:
nice!